### DISCUSSION TITLE
Dependency Graph should be displayed as a dependency tree

### DISCUSSION THREAD

**Discussion by liaodaniel** [community/community#118753] [Code Security]
### Select Topic Area

Product Feedback

### Body

The dependency graph provides a simplistic view of how to view dependencies within your repository. To help users resolve vulnerabilities in transitive dependencies, it would be beneficial if we can instead display the dependency graph more like a dependency tree i.e. show the relationship between dependencies.

Today, we have dependency snapshot submission + dependency review to report on vulnerability findings to handle the scenario of reviewing dependencies prior to merge. However, should a pull request be merged, users are left with numerous dependabot alerts raised that have no correlation back to the direct dependency. This forces users to rely on various tools to analyse the dependency tree and figure out which direct dependency they needed to remediate.

**Comment by vincerubinetti** [comment]
I'd add to this that it'd be nice to have a way to distinguish between direct and indirect/transitive dependencies, _in the API_ or exported SBOM json. Looking through every field in the SBOM, I couldn't find anything that would allow me to discriminate between them.

It seems like I'd have to scan the repo contents for the presence of each package in the main `package.json` (or equivalent) file.  But of course, that's just reimplementing what GitHub is already doing under the hood, but worse and more brittle. And it's not feasible for me, who's writing a script to analyze many repositories that are potentially in [many different languages](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-the-dependency-graph#supported-package-ecosystems).

It seems like the "Insights" tab in the website UI already has access to this information: `Detected automatically on Jul 20, 2023 (npm) · package.json`. If that could be exposed in the SBOM API response, that would be great.

---

**EDIT** I didn't realize that the [GraphQL API could do this](https://stackoverflow.com/questions/75136112/fetch-dependencies-from-a-project-in-github-using-cli):

```js
octokit.graphql(
    `
      query dependencyTree($owner: String!, $repo: String!) {
        repository(owner: $owner, name: $repo) {
          dependencyGraphManifests {
            nodes {
              blobPath
              dependencies {
                nodes {
                  packageName
                  packageManager
                  requirements
                  hasDependencies
                }
              }
            }
            edges {
              node {
                blobPath
                dependencies {
                  totalCount
                  nodes {
                    packageName
                    packageManager
                    requirements
                    hasDependencies
                  }
                }
              }
            }
          }
        }
      }
      `,
    { owner, repo: name },
  );
```

Still, it'd be good to have this in JSON you get from the website SBOM download button, and it'd be good to have feature parity between the rest and graphql APIs...
